sw/device/lib/testing/usb_testutils.c

```diff
@@ -11,39 +11,75 @@
 
 #define USBDEV_BASE_ADDR TOP_EARLGREY_USBDEV_BASE_ADDR
 
+// Specify the current file for function point reporting
+#define USBUTILS_FUNCPT_FILE USBUTILS_FUNCPT_FILE_USB_TESTUTILS
+
+#undef USBUTILS_FUNCPT
+#define USBUTILS_FUNCPT(a, b)
+
 static dif_usbdev_t usbdev;
 static dif_usbdev_buffer_pool_t buffer_pool;
 
 void usb_testutils_poll(usb_testutils_ctx_t *ctx) {
   uint32_t istate;
+
+  // Collect a set of interrupts
   CHECK_DIF_OK(dif_usbdev_irq_get_state(ctx->dev, &istate));
 
-  // Do this first to keep things going
-  CHECK_DIF_OK(dif_usbdev_fill_available_fifo(ctx->dev, ctx->buffer_pool));
+  if (!istate) {
+    // Nothing new to do right now, but keep buffers available for reception
+    CHECK_DIF_OK(dif_usbdev_fill_available_fifo(ctx->dev, ctx->buffer_pool));
+    return;
+  }
+
+  USBUTILS_FUNCPT(0x9011, istate);
+
+  // Record bus frame
+  if ((istate & (1u << kDifUsbdevIrqFrame))) {
+    // The first bus frame is 1
+    CHECK_DIF_OK(dif_usbdev_status_get_frame(ctx->dev, &ctx->frame));
+    USBUTILS_FUNCPT(0xF7AE, ctx->frame);
+    ctx->got_frame = true;
+  }
 
   // Process IN completions first so we get the fact that send completed
-  // before processing a response
+  // before processing a response to that transmission
+  // This is also important for device IN performance
   if (istate & (1u << kDifUsbdevIrqPktSent)) {
     uint16_t sentep;
     CHECK_DIF_OK(dif_usbdev_get_tx_sent(ctx->dev, &sentep));
+    USBUTILS_FUNCPT(0x5E97, sentep);
     TRC_C('a' + sentep);
-    for (int ep = 0; ep < USBDEV_NUM_ENDPOINTS; ep++) {
+    for (unsigned ep = 0; ep < USBDEV_NUM_ENDPOINTS; ep++) {
       if (sentep & (1 << ep)) {
         // Free up the buffer and optionally callback
         CHECK_DIF_OK(
             dif_usbdev_clear_tx_status(ctx->dev, ctx->buffer_pool, ep));
+
+        // TODO - supply another buffer immediately from a queue, to lessen the
```

Comment:
Indeed, mostly they're apologies for 'work not yet done because it would make the diff larger'.... far reaching changes to make the DPI signal errors to the test software at run time, for example. I'll create issues for the two or three extensive DPI changes still required/planned; that will encompass most of the 'TODO' breadcrumbs I've left